drivers/media/i2c/mira220.c

```diff
@@ -0,0 +1,2087 @@
+
+	mira220->hblank =
+		v4l2_ctrl_new_std(ctrl_hdlr, &mira220_ctrl_ops, V4L2_CID_HBLANK,
+				  mira220->mode->hblank, mira220->mode->hblank,
+				  1, mira220->mode->hblank);
+
+	// Exposure is indicated in number of lines here
+	// Max is determined by vblank + vsize and Tglob.
+	max_exposure = mira220_calculate_max_exposure_time(
+		mira220->mode->height, mira220->vblank->val,
+		mira220->mode->row_length);
+
+	mira220->exposure =
+		v4l2_ctrl_new_std(ctrl_hdlr, &mira220_ctrl_ops,
+				  V4L2_CID_EXPOSURE, MIRA220_EXPOSURE_MIN,
+				  max_exposure, 1, MIRA220_DEFAULT_EXPOSURE);
+
+	mira220->gain = v4l2_ctrl_new_std(
+		ctrl_hdlr, &mira220_ctrl_ops, V4L2_CID_ANALOGUE_GAIN,
+		MIRA220_ANALOG_GAIN_MIN, MIRA220_ANALOG_GAIN_MAX,
+		MIRA220_ANALOG_GAIN_STEP, MIRA220_ANALOG_GAIN_DEFAULT);
+
+	mira220->hflip = v4l2_ctrl_new_std(ctrl_hdlr, &mira220_ctrl_ops,
+					   V4L2_CID_HFLIP, 0, 1, 1, 0);
+	if (mira220->hflip)
+		mira220->hflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
+
+
+	mira220->vflip = v4l2_ctrl_new_std(ctrl_hdlr, &mira220_ctrl_ops,
+					   V4L2_CID_VFLIP, 0, 1, 1, 0);
+	if (mira220->vflip)
+		mira220->vflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
+
+	v4l2_ctrl_new_std_menu_items(ctrl_hdlr, &mira220_ctrl_ops,
+				     V4L2_CID_TEST_PATTERN,
+				     ARRAY_SIZE(mira220_test_pattern_menu) - 1,
+				     0, 0, mira220_test_pattern_menu);
+
+	if (ctrl_hdlr->error) {
+		ret = ctrl_hdlr->error;
+		dev_err(&client->dev, "%s control init failed (%d)\n", __func__,
+			ret);
+		goto error;
+	}
+
+	ret = v4l2_fwnode_device_parse(&client->dev, &props);
+	if (ret)
+		goto error;
+
+	ret = v4l2_ctrl_new_fwnode_properties(ctrl_hdlr, &mira220_ctrl_ops,
+					      &props);
+	if (ret)
+		goto error;
+
+	mira220->sd.ctrl_handler = ctrl_hdlr;
+
+	return 0;
+
+error:
+	v4l2_ctrl_handler_free(ctrl_hdlr);
+	mutex_destroy(&mira220->mutex);
+
+	return ret;
+}
+
+static void mira220_free_controls(struct mira220 *mira220)
+{
+	v4l2_ctrl_handler_free(mira220->sd.ctrl_handler);
+	mutex_destroy(&mira220->mutex);
+}
+
+static int mira220_probe(struct i2c_client *client)
+{
+	struct device *dev = &client->dev;
+	struct mira220 *mira220;
+	int ret;
+
+
+	mira220 = devm_kzalloc(&client->dev, sizeof(*mira220), GFP_KERNEL);
+	if (!mira220)
+		return -ENOMEM;
+
+	v4l2_i2c_subdev_init(&mira220->sd, client, &mira220_subdev_ops);
+	mira220->regmap = devm_cci_regmap_init_i2c(client, 16);
+
+	/* Get system clock (xclk) */
+	mira220->xclk = devm_clk_get(dev, NULL);
+	if (IS_ERR(mira220->xclk)) {
+		dev_err(dev, "failed to get xclk\n");
+		return PTR_ERR(mira220->xclk);
+	}
+
+	mira220->xclk_freq = clk_get_rate(mira220->xclk);
+	if (mira220->xclk_freq != MIRA220_SUPPORTED_XCLK_FREQ) {
+		dev_err(dev, "xclk frequency not supported: %d Hz\n",
+			mira220->xclk_freq);
+		return -EINVAL;
+	}
+
+	ret = mira220_get_regulators(mira220);
+	if (ret) {
+		dev_err(dev, "failed to get regulators\n");
+		return ret;
+	}
+
+	usleep_range(1000000, 1000000 + 100);
```

Comment:
Slightly crazy range of 100us tolerance on 1second.

`fsleep` is the normally recommended call now - https://www.kernel.org/doc/html/latest/timers/delay_sleep_functions.html